### PR TITLE
ReturnTypeFromStrictTernaryRector: Support complex ternaries

### PR DESCRIPTION
**Pull request by staabm**
idea is to allow ReturnTypeFromStrictTernaryRector to infer the native return type on complex typed expressions

### CODE REVIEW

File: rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/short_ternary.php.inc [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+?>
+-----
+<?php
+
+namespace Rector\Tests\TypeDeclaration\Rector\Class_\ReturnTypeFromStrictTernaryRector\Fixture;
+
+final class ShortTernary
+{
+    public function getValue(): ?\Rector\Tests\TypeDeclaration\Rector\Class_\ReturnTypeFromStrictTernaryRector\Fixture\ShortTernary
```
Comment (staabm) [inline rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/short_ternary.php.inc]: not sure why this prints the FQCN.. but I guess its intended and should be fixed by other rectors or a CS tool

File: rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/skip_phpdocs.php.inc [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace Rector\Tests\TypeDeclaration\Rector\Class_\ReturnTypeFromStrictTernaryRector\Fixture;
+
+final class SkipPhpdocs
```
Comment (staabm) [inline rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/skip_phpdocs.php.inc:5]: this test fails to skip right now. If I understood the "strict" rules correctly, they should not rely on phpdoc, right?

if so, any idea how to achieve that?

File: rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/short_ternary.php.inc (hunk 1 above)
Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/short_ternary.php.inc]: that's expected, `$rectorConfig->importNames()` can cover  that

File: rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php [hunk 3]
```diff
@@ -68,119 +73,52 @@ public function getValue($number): int
      */
     public function getNodeTypes(): array
     {
-        return [Class_::class];
+        return [ClassMethod::class, Function_::class, Closure::class];
```
Comment (staabm) [inline rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php:79]: inspired by other "ReturnTypeFromStrict" rules, just added Function/Closure support

File: rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/skip_phpdocs.php.inc (hunk 2 above)
Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/skip_phpdocs.php.inc:5]: You can compare type vs native type, like in ExprAnalyzer:

https://github.com/rectorphp/rector-src/blob/f056974ae2a1452c7cbf38f7ddcbdd29f8f21b89/src/NodeAnalyzer/ExprAnalyzer.php#L42-L52

File: rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php [hunk 4]
```diff
@@ -68,119 +73,52 @@ public function getValue($number): int
...
     public function refactorWithScope(Node $node, Scope $scope): ?Node
     {
-        $hasChanged = false;
-
-        foreach ($node->getMethods() as $classMethod) {
-            if ($classMethod->returnType instanceof Node) {
-                continue;
-            }
-
-            $onlyStmt = $classMethod->stmts[0] ?? null;
```
Comment (staabm) [inline rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php:86]: before this PR this rector only supported call-methods with a single line.
I wided the scope to call-likes with a single return

File: rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/skip_phpdocs.php.inc (hunk 2 above)
Comment (staabm) [reply rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/skip_phpdocs.php.inc:5]: thanks. I added a native type check and now I am wondering why `ternary_var_call.php.inc` is failling.
rector type resolving tells me the native type of the ternary is `MixedType` but running the same code directly in PHPStan returns a correct native type..

Comment (staabm) [reply rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/skip_phpdocs.php.inc:5]: I dropped the failling test for now, as I don't see it important enough to block the actual useful change

**Comment by staabm** [comment]
should be ready for review/merge @samsonasik

File: rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php [hunk 5]
```diff
@@ -68,119 +76,71 @@ public function getValue($number): int
...
 
-        return null;
+        $node->returnType = $returnTypeNode;
+        return $node;
     }
 
     public function provideMinPhpVersion(): int
     {
-        return PhpVersionFeature::SCALAR_TYPES;
+        return PhpVersionFeature::NULLABLE_TYPE;
```
Comment (samsonasik) [inline rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php]: Php 7.0 for scalar feature should be ok, as far as return type validated via StaticTypeMapper which check feature type

Comment (staabm) [reply rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php]: done :+1:

**Review by samsonasik — APPROVED** [review]

**Comment by TomasVotruba** [comment]
Looks good, thank you :+1:

**Pull request merged**